core/chains/evm/txm/txm.go

```diff
@@ -0,0 +1,447 @@
+	config          Config
+	metrics         *txmMetrics
+
+	nonceMapMu sync.Mutex
+	nonceMap   map[common.Address]uint64
+
+	triggerCh map[common.Address]chan struct{}
+	stopCh    services.StopChan
+	wg        sync.WaitGroup
+}
+
+func NewTxm(lggr logger.Logger, chainID *big.Int, client Client, attemptBuilder AttemptBuilder, txStore TxStore, stuckTxDetector StuckTxDetector, config Config, keystore Keystore) *Txm {
+	return &Txm{
+		lggr:            logger.Sugared(logger.Named(lggr, "Txm")),
+		keystore:        keystore,
+		chainID:         chainID,
+		client:          client,
+		attemptBuilder:  attemptBuilder,
+		txStore:         txStore,
+		stuckTxDetector: stuckTxDetector,
+		config:          config,
+		nonceMap:        make(map[common.Address]uint64),
+		triggerCh:       make(map[common.Address]chan struct{}),
+	}
+}
+
+func (t *Txm) Start(ctx context.Context) error {
+	return t.StartOnce("Txm", func() error {
+		tm, err := NewTxmMetrics(t.chainID)
+		if err != nil {
+			return err
+		}
+		t.metrics = tm
+		t.stopCh = make(chan struct{})
+
+		addresses, err := t.keystore.EnabledAddressesForChain(ctx, t.chainID)
+		if err != nil {
+			return err
+		}
+		for _, address := range addresses {
+			t.startAddress(address)
+		}
+		return nil
+	})
+}
+
+func (t *Txm) startAddress(address common.Address) {
+	triggerCh := make(chan struct{}, 1)
+	t.triggerCh[address] = triggerCh
+
+	t.wg.Add(2)
+	go t.broadcastLoop(address, triggerCh)
+	go t.backfillLoop(address)
+}
+
+func (t *Txm) initializeNonce(ctx context.Context, address common.Address) {
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, pendingNonceDefaultTimeout)
+	defer cancel()
+	for {
+		pendingNonce, err := t.client.PendingNonceAt(ctxWithTimeout, address)
+		if err != nil {
+			t.lggr.Errorw("Error when fetching initial nonce", "address", address, "err", err)
+			select {
+			case <-time.After(pendingNonceRecheckInterval):
+			case <-ctx.Done():
+				t.lggr.Errorw("context error", "err", context.Cause(ctx))
+				return
+			}
+			continue
+		}
+		t.setNonce(address, pendingNonce)
+		t.lggr.Debugf("Set initial nonce for address: %v to %d", address, pendingNonce)
+		return
+	}
+}
+
+func (t *Txm) Close() error {
+	return t.StopOnce("Txm", func() error {
+		close(t.stopCh)
+		t.wg.Wait()
+		return nil
+	})
+}
+
+func (t *Txm) HealthReport() map[string]error {
+	return map[string]error{t.lggr.Name(): t.Healthy()}
+}
+
+func (t *Txm) CreateTransaction(ctx context.Context, txRequest *types.TxRequest) (tx *types.Transaction, err error) {
+	tx, err = t.txStore.CreateTransaction(ctx, txRequest)
+	if err == nil {
+		t.lggr.Infow("Created transaction", "tx", tx)
+	}
+	return
+}
+
+func (t *Txm) Trigger(address common.Address) {
+	if !t.IfStarted(func() {
+		triggerCh, exists := t.triggerCh[address]
+		if !exists {
+			return
+		}
+		triggerCh <- struct{}{}
+	}) {
+		t.lggr.Error("Txm unstarted")
+	}
+}
+
+func (t *Txm) Abandon(address common.Address) error {
+	// TODO: restart txm
```

Comment:
I assume this isn't needed before merge?

Reply:
Yes, but I already have a pending PR that addresses this.